data/posts/react-router-and-react-server-components.md

```diff
@@ -0,0 +1,149 @@
+
+### Streaming UI
+
+**Framework Mode**: Return promises from loaders and use the `<Await>` component.
+
+```js
+// Framework Mode
+export function loader() {
+  return defer({
+    criticalData: getCriticalData(),
+    slowData: getSlowData(), // This is a promise
+  });
+}
+
+export default function Page() {
+  const { criticalData, slowData } = useLoaderData();
+  return (
+    <>
+      <h1>{criticalData.title}</h1>
+      <Suspense fallback={<Spinner />}>
+        <Await resolve={slowData}>
+          {(data) => <SlowContent data={data} />}
+        </Await>
+      </Suspense>
+    </>
+  );
+}
+```
+
+**React RSC**: Pass promises as props and use `await` on the server and `use(promise)` on the client.
+
+**React Router RSC**: Again, both patterns are available to you.
+
+### Code Splitting Routes
+
+**Framework Mode**: A `routes.ts` config file processed by a bundler plugin generates a manifest that maps routes to chunks.
+
+**React RSC**: Dynamic imports on the server combined with `"use client"` directives mean the browser only downloads what it needs.
+
+## Under the Hood: How Framework Mode Works
+
+React Router's Framework Mode is powered by some clever bundler integration:
+
+1. You define routes in a `routes.ts` config file
+2. A bundler plugin reads this config and generates:
+   - A manifest mapping routes to chunks for the browser
+   - A server entry containing the manifest and server-side route mappings
+3. The React Router runtime uses these generated files to handle routing
+
+This approach has served us well, but it requires significant bundler integration and complexity.
+
+## The New Way: React Router RSC
+
+React Router RSC takes a different approach that's actually simpler:
+
+- Routes defined in app code like any other data structure (no more special `routes.ts` bundler plugins or build-time generation required)
+- A function that handles routing HTML vs data requests
+- A function that generates RSC payloads that get rendered to HTML somewhere
+
+Notice that these are now just plain data structures and function calls. These APIs replace what previously required an entire bundler plugin and manifest generation system. The result? A simpler, more direct architecture that allows React Router to be a library again for complex full stack apps, only possible thanks to RSC.
+
+It's worth calling out that this simplification is useful _even if you're not using server components_. Even if all of your routes are client components, as they are today, this new architecture means that advanced usage of React Router will no longer require a bundler-specific Framework Mode plugin.
+
+## What This Means for You
+
+- **If you're using React Router in Framework Mode:** Our intention is for the future of Framework Mode to be built on top of React Router RSC, but your existing code will continue to work. When the transition happens, it will be seamless from your perspective. _No migration necessary!_
```

Comment:
```suggestion
- **If you're using React Router in Framework Mode:** Our intention is for the future of Framework Mode to be built on top of React Router RSC, so your existing code will continue to work. When the transition happens, it will be seamless from your perspective. _No migration necessary!_
```